fred2/sexp_tree.cpp

```diff
@@ -1944,33 +1914,141 @@ SCP_string sexp_tree::match_closest_operator(const SCP_string &str, int node)
+		dc->SelectObject(GetFont());
+		m_font_max_width = 0;
+		for (auto& op : Operators)
+		{
+			auto font_extent = dc->GetTextExtent(op.text.c_str());
+			if (font_extent.cx > m_font_max_width)
+				m_font_max_width = font_extent.cx;
 		}
+
+		// adjust for scroll bar and border edge
+		m_font_max_width += GetSystemMetrics(SM_CXVSCROLL) + 2 * GetSystemMetrics(SM_CXEDGE);
 	}
 
-	Assert(best >= 0);  // we better have some valid operator at this point.
-	return Operators[best].text;
+	// calculate position and size of the dropdown
+	RECT item_rect, dropdown_rect;
+	GetItemRect(h, &item_rect, TRUE);
+	dropdown_rect.top = item_rect.top;
+	dropdown_rect.left = item_rect.left;
+	dropdown_rect.right = dropdown_rect.left + m_font_max_width;
+	dropdown_rect.bottom = dropdown_rect.top + m_font_height * 10;
+
+	// create or just position it
+	if (!m_operator_popup_created)
+	{
+		m_operator_box.Create(WS_CHILD | WS_VISIBLE | WS_VSCROLL | CBS_SIMPLE | CBS_HASSTRINGS | CBS_OWNERDRAWFIXED, dropdown_rect, this, IDC_SEXP_POPUP_LIST);
+		m_operator_box.SetFont(GetFont());
+		m_operator_popup_created = true;
+	}
+	else
+	{
+		m_operator_box.MoveWindow(&dropdown_rect);
+	}
+
+	m_operator_box.refresh_popup_operators(opf_type, tree_nodes[item_index].text);
+
+	m_operator_box.ShowWindow(SW_SHOWNORMAL);
+	m_operator_box.SetFocus();
+	m_operator_popup_active = true;
+}
+
+void sexp_tree::end_operator_edit(bool confirm)
+{
+	if (!m_operator_popup_active)
+		return;
+
+	m_operator_box.cleanup(confirm);
+	m_operator_box.ShowWindow(SW_HIDE);
+	m_operator_popup_active = false;
 }
 
 // this really only handles messages generated by the right click popup menu
+// now it also handles messages from the operator combo box
 BOOL sexp_tree::OnCommand(WPARAM wParam, LPARAM lParam)
 {
-	int i, z, id, node, op;
+	int i, z, id, data, node, op;
 	sexp_list_item *list, *ptr;
 	HTREEITEM h;
 
-	if ((item_index >= 0) && (item_index < total_nodes))
+	if ((item_index >= 0) && (item_index < total_nodes) && !tree_nodes.empty())
 		item_handle = tree_nodes[item_index].handle;
 
 	id = LOWORD(wParam);
+	data = HIWORD(wParam);
+
+#ifndef NDEBUG
+	// Sanity check: None of these #defines used in this function should overlap with any SEXP operators, or mysterious bugs will ensue
+	// note: this won't catch operator values plus OP_REPLACE_FLAG or OP_INSERT_FLAG
+	switch (id)
+	{
+		case ID_SEXP_TREE_ADD_VARIABLE:
+		case ID_SEXP_TREE_MODIFY_VARIABLE:
+		case ID_EDIT_SEXP_TREE_EDIT_CONTAINERS:
+		case ID_EDIT_COPY:
+		case ID_EDIT_PASTE:
+		case ID_EDIT_PASTE_SPECIAL:
+		case ID_SPLIT_LINE:
+		case ID_EXPAND_ALL:
+		case ID_EDIT_TEXT:
+		case ID_EDIT_COMMENT:
+		case ID_EDIT_BG_COLOR:
+		case ID_REPLACE_NUMBER:
+		case ID_REPLACE_STRING:
+		case ID_ADD_STRING:
+		case ID_ADD_NUMBER:
+		case ID_EDIT_CUT:
+		case ID_DELETE:
+		case IDC_SEXP_POPUP_LIST:
+			if (id >= FIRST_OP && id < sexp::operator_upper_bound())
+				Error(LOCATION, "A resource definition (%d) overlaps with an operator value!", id);
```

Comment:
Since this is a debug-only error, and a static coding error and not a runtime data error, it should be erroring using Assert or unreachable